rhtconsulting/rhc-ose
# Add OpenStack terminate

#### What does this PR do?

This PR adds a playbook that relies on your sourced OpenStack credentials to get a list of nova instances matching a supplied **env_id**, then unregister from subscripiton-manager, delete these instances, then delete any attached volumes. This is useful when provisioning and tearing down environments using the provisioning playbooks.

Currently this uses environment variables to list instances. Only the instances matching the supplied string will be deleted, so this has been safely tested to not damage any non-matching instances. However, as with any string matching, this should be as specific and unique as possible. An 'env_id' of "testenv" will also match "testenv1" and "testenv2" and so on. 
#### How should this be manually tested?

To test this first source your openrc.sh file, provision an environment with a unique 'env_id', optionally list instances and volumes (to verify after), then run the playbook. Verify that everything was cleaned up after.

```
source ~/.openstack/openrc.sh
nova list
nova volume-list
ansible-playbook openstack-deprovision.yml -e 'env_id=testenvX'
nova list
nova volume-list
```

Additional items that can be tested:
- No successful connection to environment should halt with an error
- No specified **env_id** should fail
- **env_id** length of less than 8 should fail
- More than 6 matching instances should fail
- Overriding the above should succeed either by individual variable override or via **really_really_sure=true** parameter
- Override with no specified **env_id** should match all instances
- 'dry_run' shows what would be deleted, this works well as the actual tasks will be skipped by show the IDs that would be deleted when set to true
- Debug info is available for the custom host group if **dry_run=true** is set
- If images used in matching instances are not unique, a warning will prompt. The reason is that if a different **ansible_ssh_user** is needed for the different users, then the subscription-manager task might fail on instances with different users
#### Is there a relevant Issue open for this?

No current issues.
#### Who would you like to review this?

/cc @oybed @sabre1041 @etsauer
